### ISSUE TITLE
how to deliver futharked libs?

### ISSUE THREAD

**Issue by arkanoid87** [PMunch/futhark#7]
Thanks to Futhark wrapping is easy and I've successfully produced another (full toy) wrapping with working example.

I'll link it here just to prove it, but the question really is if futhark-generated bindings should be delivered or not, and if yes, how?
- should we deliver futhark_*.nim tagged to lib version?
- should we deliver project with importc block?
- is making packages for plain futharked libs a thing?
This should be a topic of discussion, but not sure if the right place is here or on the forum.

Here's a small example wrapping `libgirepository` that calls GLib,base64_encode via GObject interface
https://gi.readthedocs.io

C Example
```C
#include <girepository.h>

int main(void)
{
    GIRepository *repository;
    GError *error = NULL;
    GIBaseInfo *base_info;
    GIArgument in_args[2];
    GIArgument retval;

    repository = g_irepository_get_default();
    g_irepository_require(repository, "GLib", "2.0", 0, &error);
    if (error)
    {
        g_error("ERROR: %s\n", error->message);
        return 1;
    }

    base_info = g_irepository_find_by_name(repository, "GLib", "base64_encode");
    if (!base_info)
    {
        g_error("ERROR: %s\n", "Could not find GLib.base64_encode");
        return 1;
    }

    in_args[0].v_pointer = (gpointer) "Hello World!";
    in_args[1].v_ulong = 12;

    if (!g_function_info_invoke((GIFunctionInfo *)base_info,
                                (const GIArgument *)&in_args,
                                2,
                                NULL,
                                0,
                                &retval,
                                &error))
    {
        g_error("ERROR: %s\n", error->message);
        return 1;
    }

    g_print("%s\n", retval.v_string);

    g_base_info_unref(base_info);

    return 0;
}
```

Nim + Futhark
```Nim
import futhark

importc:
  absPath "/usr/lib/clang/10/include"
  path "/usr/include/gobject-introspection-1.0"
  path "/usr/include/glib-2.0"
  path "/usr/lib/x86_64-linux-gnu/glib-2.0/include"
  path "/usr/include/glib-2.0/glib"
  "girepository.h"
  "gitypes.h"
  "gerror.h"
  "gifunctioninfo.h"
  "gibaseinfo.h"


# https://gi.readthedocs.io/en/latest/writingbindings/libgirepository.html

converter toGchar(s: string): ptr gchar =
  cast[ptr gchar](s.cstring)

converter `$`(s: ptr gchar): string =
  $cast[cstring](s)


var
  repository: ptr GIRepository
  error: ptr GError
  base_info: ptr GIBaseInfo
  in_args: array[2, GIArgument]
  retval: GIArgument

repository = g_irepository_get_default()
discard g_irepository_require(
  repository,
  "GLib",
  "2.0",
  Girepositoryloadflaglazy,
  error.addr)
if error != nil:
  echo("ERROR: " & $error[].message)
  quit 1

base_info = g_irepository_find_by_name(
  repository,
  "GLib",
  "base64_encode")
if base_info == nil:
  echo "ERROR: Could not find GLib.base64_encode"
  quit 1

let text2encode = "Hello World!"

in_args[0].v_pointer = cast[gpointer](text2encode.cstring)
in_args[1].v_int = text2encode.len.cint

let invoked = g_function_info_invoke(
  cast[ptr GIFunctionInfo](base_info),
  cast[ptr GIArgument](in_args.addr),
  2.cint,
  nil,
  0.cint,
  retval.addr,
  error.addr)

echo retval.v_string

if invoked == 0:
  echo "ERROR: " & $error[].message
  quit 1

g_base_info_unref(base_info)
```

config.nims
```nim
--maxLoopIterationsVM: "100000000"
--passL: "-lgirepository-1.0"
```

Output for both C and Nim
```bash
SGVsbG8gV29ybGQh
```

**Comment by PMunch** [comment]
This is an interesting question. The intention behind Futhark was that you shouldn't need wrappers, and therefore you shouldn't need to create packages for Futhark-wrapped projects. Futhark also executes macros and such from the C code, so if you where to distribute a Futhark-wrapping it would definitely be best to do it as a `importc` block. This ensures that the wrapping are actually intended for the system it's building on, and allows the user to reconfigure the C library and have the wrapping still be correct. This is something which isn't really discussed in the other wrapping tools, but many C libraries have compile-time switches that will change what is actually defined in the C sources.

That being said I can see how it would sometimes be beneficial to be able to generate stand-alone wrappers. I have planned a mode for Futhark that would simply not generate all the `when defined` logic it currently uses which would make the results more apt for post-processing. This should probably be the way you want to do it if you're creating bindings and don't want to just have an `importc` with the low-level -> high-level binding code as your actual project.

**Comment by PMunch** [comment]
So a little update on this based on what I've been doing with my wrapped libraries. I still stand by the core idea that Futhark should be used to wrap libraries on the users machine to ensure the bindings are actually correct. However in my own projects I've opted for a `-d:useFuthark` switch which either imports Futhark and calls `importc` or simply includes a pre-generated `futhark_*.nim` file. This allows people to use my library without having to install and set up Futhark and Clang, and allows slower targets like a Raspberry Pi for example to not have to do the conversion. It also allows users who want to enable a feature through C `#define` switches to do so without having to clone and mess with the library manually.

I've also recently included a `-d:nodeclguards` which makes the generated files much easier to read and post-process, so if you opt to go that route this should now be easier and more convenient.
